pkg/controller/util/endpoint/controller_utils_test.go

```diff
@@ -224,3 +231,297 @@ func TestShouldPodBeInEndpoints(t *testing.T) {
+	tests := []struct {
+		key      string
+		name     string
+		selector map[string]string
+		action   Action
+		exist    bool
+		expect   labels.Selector
+	}{
+		// test cases are dependent because update/delete case will change the ServiceSelectorCache.
+		{
+			key:    "test/service-0",
+			name:   "get a nonexistent selector",
+			action: Get,
+			exist:  false,
+		},
+		{
+			key:      "test/service-0",
+			name:     "add service-0 selector",
+			selector: map[string]string{"app": "test-0"},
+			action:   Update,
+			expect:   labels.Set(map[string]string{"app": "test-0"}).AsSelectorPreValidated(),
+		},
+		{
+			key:    "test/service-0",
+			name:   "get exist selector",
+			action: Get,
+			exist:  true,
+			expect: labels.Set(map[string]string{"app": "test-0"}).AsSelectorPreValidated(),
+		},
+		{
+			key:    "test/service-0",
+			name:   "delete service-0",
+			action: Delete,
+		},
+		{
+			key:      "test/service-0",
+			name:     "update service-0 selector",
+			selector: map[string]string{"app": "test-0", "app1": "test-1"},
+			action:   Update,
+			expect:   labels.Set(map[string]string{"app": "test-0", "app1": "test-1"}).AsSelectorPreValidated(),
+		},
+		{
+			key:    "test/service-0",
+			name:   "get service-0 selector after updated",
+			action: Get,
+			exist:  true,
+			expect: labels.Set(map[string]string{"app": "test-0", "app1": "test-1"}).AsSelectorPreValidated(),
+		},
+		{
+			key:      "test/service-1",
+			name:     "add service-1 selector",
+			selector: map[string]string{"app": "test-1"},
+			action:   Update,
+			expect:   labels.Set(map[string]string{"app": "test-1"}).AsSelectorPreValidated(),
+		},
+		{
+			key:    "test/service-1",
+			name:   "get service-1 selector",
+			action: Get,
+			exist:  true,
+			expect: labels.Set(map[string]string{"app": "test-1"}).AsSelectorPreValidated(),
+		},
+		{
+			key:    "test/service-1",
+			name:   "delete service-1 selector",
+			action: Delete,
+		},
+		{
+			key:    "test/service-1",
+			name:   "get service-1 after deleted",
+			action: Get,
+			exist:  false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			switch test.action {
+			case Get:
+				selector, ok := cache.Get(test.key)
+				if ok != test.exist {
+					t.Errorf("%s failed with key existence %t but expect %t", test.name, ok, test.exist)
+				}
+				if test.exist == true && !reflect.DeepEqual(selector, test.expect) {
+					t.Errorf("%s failed with key existence %t but expect %t, selector %v but expect %v", test.name, ok, test.exist, selector, test.expect)
+				}
+			case Update:
+				if selector := cache.Update(test.key, test.selector); !reflect.DeepEqual(selector, test.expect) {
+					t.Errorf("%s failed with selector %v but expect %v", test.name, selector, test.expect)
+				}
+			case Delete:
+				cache.Delete(test.key)
+			default:
+				t.Errorf("Unknown action %v", test.action)
+			}
+		})
+	}
+}
+
+func BenchmarkGetPodServiceMemberships(b *testing.B) {
```

Comment:
@robscott I have tested the performance of GetPodServiceMemberships by BenchmarkGetPodServiceMemberships with b.N = 20000
When I set service number as 500, 1000 and 5000
time wastes(nano second):
3947460065->2539256764(64%)
7816078517->4944378261(63%)
48047946858->32854451880(68%)
each result is an average of five.

Reply:
yes, larger numbers is the time calling the original GetPodServiceMemberships function and the newer numbers calling the new GetPodServiceMemberships function.

Reply:
just make original GetPodServiceMemberships become a method of ServiceSelectorCache without any change in the function body.

Reply:
Nice! That's a great improvement.

Reply:
seems get function should also hold the lock, performance will decline 1%~2% after get function hold the read lock.